File: compiler/bench/binsearch.fs
\ expected: 499500000
\ Binary search benchmark - search 1M times in sorted array

1000 constant SIZE
1000000 constant ITERS
create arr SIZE cells allot

: arr@ ( i -- n ) cells arr + @ ;
: arr! ( n i -- ) cells arr + ! ;

: bsearch ( val -- idx|-1 )
  0 SIZE 1-             \ val lo hi
  begin 2dup <= while
    2dup + 2 /          \ val lo hi mid
    dup arr@            \ val lo hi mid arr[mid]
    4 pick = if         \ found
      nip nip nip exit
    then
    dup arr@ 4 pick < if
      \ arr[mid] < val, search right half
      rot drop 1+ swap  \ val mid+1 hi ; lo=mid+1
    else
      \ arr[mid] > val, search left half
      nip 1- swap       \ val lo mid-1 ; hi=mid-1
    then
  repeat
  2drop drop -1 ;

: init-arr ( -- )
  SIZE 0 do i i arr! loop ;

: bench ( -- sum )
  0
  ITERS 0 do
    i SIZE mod bsearch +
  loop ;

: main
  init-arr
  bench . cr ;

File: compiler/bench/binsearch.c
// expected: 499500000
// Binary search benchmark - search 1M times in sorted array

#include <stdio.h>

#define SIZE 1000
#define ITERS 1000000

int arr[SIZE];

int bsearch_impl(int val) {
    int lo = 0, hi = SIZE - 1;
    while (lo <= hi) {
        int mid = (lo + hi) / 2;
        if (arr[mid] == val) return mid;
        if (arr[mid] < val) {
            lo = mid + 1;
        } else {
            hi = mid - 1;
        }
    }
    return -1;
}

void init_arr(void) {
    for (int i = 0; i < SIZE; i++) {
        arr[i] = i;
    }
}

long bench(void) {
    long sum = 0;
    for (int i = 0; i < ITERS; i++) {
        sum += bsearch_impl(i % SIZE);
    }
    return sum;
}

int main(void) {
    init_arr();
    printf("%ld\n", bench());
    return 0;
}
